src/spark_rapids_ml/knn.py

```diff
@@ -235,54 +210,60 @@ def setK(self, value: int) -> "NearestNeighbors":
     def _create_pyspark_model(self, result: Row) -> "NearestNeighborsModel":
         return NearestNeighborsModel.from_row(result)
 
-    def fit(self, dataset: DataFrame, params: Optional[Dict[Param[Any], Any]] = None) -> "NearestNeighbors":  # type: ignore
+    def _fit(self, dataset: DataFrame) -> "NearestNeighborsModel":
         self.item_df = dataset
         if not self.isDefined("id_col"):
-            self.item_df = self._df_zip_with_index(self.item_df)
+            self.item_df = self._df_zip_with_index(self.item_df, self.getIdCol())
 
         self.processed_item_df = self.item_df.withColumn(
             alias.label, lit(self.label_isdata)
         )
-        return self
-
-    def kneighbors(self, query_df: DataFrame) -> Tuple[DataFrame, DataFrame, DataFrame]:
-        query_default_num_partitions = query_df.rdd.getNumPartitions()
-        if not self.isDefined("id_col"):
-            query_df = self._df_zip_with_index(query_df)
-        processed_query_df = query_df.withColumn(alias.label, lit(self.label_isquery))
 
-        union_df = self.processed_item_df.union(processed_query_df)
-
-        pipelinedrdd = self._fit(union_df)
-        pipelinedrdd = pipelinedrdd.repartition(query_default_num_partitions)  # type: ignore
-        knn_rdd = pipelinedrdd.flatMap(
-            lambda row: list(zip(row["query_id"], row["indices"], row["distances"]))
+        model = self._create_pyspark_model(
+            Row(
+                item_df=self.item_df,
+                processed_item_df=self.processed_item_df,
+                label_isdata=self.label_isdata,
+                label_isquery=self.label_isquery,
+            )
```

Comment:
We should test this at scale to see if/when we hit limits.

Reply:
Yeah, marked in Jira ticket DML-457